capstone/src/main/java/com/google/sps/search/SearchPredictor.java

```diff
@@ -35,29 +39,81 @@ public SearchPredictor() {
       names.add(name);
     }
     return names;
   }
 
   private void populateTrie() {
     for (String fullName : names) {
-      String[] split = fullName.split("+");
+      String[] split = fullName.split("@");
       // not robust for internationalization
       String firstName = split[0];
       String lastName = split[1];
-      firstNameTrie.insert(firstName, fullName);
-      lastNameTrie.insert(lastName, fullName);
+      String properName = firstName + " " + lastName;
+      firstNameTrie.insert(firstName, properName);
+      lastNameTrie.insert(lastName, properName);
     }
   }
 
   // incomplete
-  public Set<String> suggest(String phrase) {
-    Set<String> names = new HashSet<String>();
-    names.addAll(firstNameTrie.searchWithPrefix(phrase, phrase));
-    names.addAll(lastNameTrie.searchWithPrefix(phrase, phrase));
-    return names;
+  public List<String> suggest(String input) {
+    Map<String, Integer> namesScore = new HashMap<String, Integer>();
+    String[] split = input.split(" ");
+    for (int i = 0; i < split.length; i++) {
+      String partialName = split[i];
+      Set<String> firstNameSuggestions = firstNameTrie.searchWithPrefix(partialName, partialName);
+      Set<String> lastNameSuggestions = lastNameTrie.searchWithPrefix(partialName, partialName);
+      // Matching prefix first name is weighted more heavily
+      this.addToMap(namesScore, firstNameSuggestions, partialName, 2);
+      this.addToMap(namesScore, lastNameSuggestions, partialName, 1);
+    }
+    List<String> sortedNames = this.sortNames(namesScore);
+    return sortedNames;
+  }
+
+  public List<String> sortNames(Map<String, Integer> namesScore) {
+    List<String> sortedNames = new ArrayList<String>();
+
+    List<Map.Entry<String, Integer>> entries =
+        new ArrayList<Map.Entry<String, Integer>>(namesScore.entrySet());
+
+    Collections.sort(
+        entries,
+        new Comparator<Map.Entry<String, Integer>>() {
+          @Override
+          public int compare(Map.Entry<String, Integer> a, Map.Entry<String, Integer> b) {
+            return Integer.compare(b.getValue(), a.getValue());
+          }
+        });
+
+    for (Map.Entry<String, Integer> entry : entries) {
+      sortedNames.add(entry.getKey());
+    }
+
+    return sortedNames;
+  }
+
+  private Map<String, Integer> addToMap(
+      Map<String, Integer> namesScore, Set<String> set, String partialName, int increment) {
+    for (String name : set) {
+      if (!namesScore.containsKey(name)) {
+        namesScore.put(name, increment);
+      } else {
+        int score = namesScore.get(name) + 1;
+        namesScore.put(name, score);
+      }
+
+      // If there is a complete name match for first or last name, increment score by 5
+      String[] split = name.split(" ");
+      partialName = partialName.toUpperCase();
+      if (partialName.equals(split[0]) || partialName.equals(split[1])) {
+        int score = namesScore.get(name) + FIVE;
```

Comment:
I just needed a number that would increase the rank if either the first or last name matched completely. I had that if the first name prefix matched it would be 2 points and if last name prefix matched it would be 1 point. For example if the input is Andrew S, then Sasha Smith could be a possible option but I want names like Andrew Sweet, Andrew Brown etc to be ranked about Sasha Smith

Reply:
Okay, makes sense!